### ISSUE TITLE
Windows Issues with "Replace error-prone file timestamp check by checksum when loading deps"

### ISSUE THREAD

**Issue by zoffixznet** [rakudo/rakudo#1662]
When I tried to install a module with recentish rakudo on Windows 10, I got a bunch of errors about "This type cannot unbox to a native string: P6opaque, Failure".

This happened with more than one module and full output is available here, on File::Which Issue I originally filed: https://github.com/azawawi/perl6-file-which/issues/9#issue-309154423

I traced the problem to `nqp::sha1($srcIO.slurp(:enc<iso-8859-1>))` call in `src/core/CompUnit/PrecompilationUnit.pm6` which looks to have been introed in https://github.com/rakudo/rakudo/commit/ca0a74398b9385053c79de62a42418258905c715#diff-151af5b717fd68282437bd98c256a260 

There's [this conditional](https://github.com/rakudo/rakudo/blob/17ca56b4ebb82d077b324bf0e64ccca0d5e09d4c/src/core/CompUnit/PrecompilationUnit.pm6#L60-L62) that looks to be incorrect, as it'd ever check `.e` if `$srcIO` was false.

```
            unless $srcIO {
                return False unless $srcIO.e;
            }
```

I suspect changing that to `return False unless $srcIO and $srcIO.e;` might fix my original issue, but I'm wondering whether that'd just be masking another issue?

I dumped the path it's trying to slurp and it's
```
   IO::Path.new("perl#sources\\51E302443A2C8FF185ABC10CA1E5520EFEE885A1", :SPEC(IO::Spec::Win32), :CWD("C:\\Temp\\perl6-file-which"))
```

Kinda looks weird. Is that what the path would supposed to look? I'm wondering if there isn't some replacement of `perl#sources` that should've occured, but didn't because the path separator is `\` and not `/`

**Comment by niner** [comment]
I'm quite sure the unless $srcIO block is wrong. return False unless $srcIO.e; should do the trick

Parsing seems to work just fine depite the backslashes:
```
perl6 -e 'say CompUnit::Repository::Spec.from-string("perl#sources\\51E302443A2C8FF185ABC10CA1E5520EFEE885A1")'
CompUnit::Repository::Spec.new(short-id => "perl", options => {}, path => "sources\\51E302443A2C8FF185ABC10CA1E5520EFEE885A1", Str => "perl#sources\\51E302443A2C8FF185ABC10CA1E5520EFEE885A1")
```

That the "perl#" part is still in $srcIO points at CompUnit::RepositoryRegistry.file-for-spec failing to find the "perl" repository which could happen if repository setup was screwed. What does $*REPO.repo-chain look like?

**Comment by zoffixznet** [comment]
> What does $*REPO.repo-chain look like?

```
C:\Temp\perl6-file-which>perl6 -e "say $*REPO.repo-chain"
(inst#C:\Users\zoffi\.perl6 inst#C:\rakudo\install\share\perl6\site inst#C:\rakudo\install\share\perl6\vendor inst#C:\rakudo\install\share\perl6 CompUnit::Repository::AbsolutePath.new(next-repo => CompUnit::Repository::NQP.new(next-repo => CompUnit::Repository::Perl5.new(next-repo => CompUnit::Repository))) CompUnit::Repository::NQP.new(next-repo => CompUnit::Repository::Perl5.new(next-repo => CompUnit::Repository)) CompUnit::Repository::Perl5.new(next-repo => CompUnit::Repository))

C:\rakudo>perl6 -MNotThere -e''
===SORRY!===
Could not find NotThere at line 1 in:
    C:\Users\zoffi\.perl6
    C:\rakudo\install\share\perl6\site
    C:\rakudo\install\share\perl6\vendor
    C:\rakudo\install\share\perl6
    CompUnit::Repository::AbsolutePath<129293040>
    CompUnit::Repository::NQP<135209224>
    CompUnit::Repository::Perl5<135209264>
```

I also inserted this code:
```
            dd [
              CompUnit::RepositoryRegistry.file-for-spec($dependency.src),
              $dependency.src.IO,
              $srcIO.e,
              $srcIO,
            ];
```

And the output was:
```
[Any, 
IO::Path.new("perl#sources\\51E302443A2C8FF185ABC10CA1E5520EFEE885A1", :SPEC(IO::Spec::Win32), :CWD("C:\\Temp\\perl6-file-which")), 
Bool::False, 
IO::Path.new("perl#sources\\51E302443A2C8FF185ABC10CA1E5520EFEE885A1", :SPEC(IO::Spec::Win32), :CWD("C:\\Temp\\perl6-file-which"))]
```

So `CompUnit::RepositoryRegistry.file-for-spec($dependency.src),` returns an undefined value and `//` fallbacks to `$dependency.src.IO,` Is `$dependency.src` already meant to have the `perl#` part removed from it by that point?

**Comment by zoffixznet** [comment]
FWIW, after I change the ` unless $srcIO` bit, I get errors about `'dynext/perl6_ops_moar.dll'`

```
C:\rakudo>zef install https://github.com/perl6/gtk-simple.git
===> Searching for missing dependencies: LWP::Simple, Shell::Command
===> Updated cpan mirror: https://raw.githubusercontent.com/ugexe/Perl6-ecosystems/master/cpan.json
===> Updated p6c mirror: http://ecosystem-api.p6c.org/projects.json
===> Searching for missing dependencies: MIME::Base64, URI, JSON::Tiny, File::Which, File::Find
===> Building: GTK::Simple:ver<0.1.7>
Unhandled exception: failed to load library 'dynext/perl6_ops_moar.dll'
   at <unknown>:1  (C:\rakudo\perl6.moarvm:<dependencies+deserialize>)
===> Building [OK] for GTK::Simple:ver<0.1.7>
===> Testing: MIME::Base64:ver<1.2.1>:auth<github:retupmoca>
Unhandled exception: failed to load library 'dynext/perl6_ops_moar.dll'
   at <unknown>:1  (C:\rakudo\perl6.moarvm:<dependencies+deserialize>)
Unhandled exception: failed to load library 'dynext/perl6_ops_moar.dll'
   at <unknown>:1  (C:\rakudo\perl6.moarvm:<dependencies+deserialize>)
Unhandled exception: failed to load library 'dynext/perl6_ops_moar.dll'
   at <unknown>:1  (C:\rakudo\perl6.moarvm:<dependencies+deserialize>)
Unhandled exception: failed to load library 'dynext/perl6_ops_moar.dll'
   at <unknown>:1  (C:\rakudo\perl6.moarvm:<dependencies+deserialize>)
===> Testing [FAIL]: MIME::Base64:ver<1.2.1>:auth<github:retupmoca>
Aborting due to test failure: MIME::Base64:ver<1.2.1>:auth<github:retupmoca> (use --force-test to override)
  in code  at C:\rakudo\install\share\perl6\site\sources\8244C3B17ACA61B0EC04857BB3283A8FAF7A186D (Zef::Client) line 374
  in method test at C:\rakudo\install\share\perl6\site\sources\8244C3B17ACA61B0EC04857BB3283A8FAF7A186D (Zef::Client) line 354
  in code  at C:\rakudo\install\share\perl6\site\sources\8244C3B17ACA61B0EC04857BB3283A8FAF7A186D (Zef::Client) line 523
  in sub  at C:\rakudo\install\share\perl6\site\sources\8244C3B17ACA61B0EC04857BB3283A8FAF7A186D (Zef::Client) line 520
  in method install at C:\rakudo\install\share\perl6\site\sources\8244C3B17ACA61B0EC04857BB3283A8FAF7A186D (Zef::Client) line 621
  in sub MAIN at C:\rakudo\install\share\perl6\site\sources\81436475BD18D66BFD96BBCEE07CCCDC0F368879 (Zef::CLI) line 152
  in block <unit> at C:\rakudo\install\share\perl6\site\resources\0F85B1755CC82172DA91B4DC34A736FA364C0D9D line 3
  in sub MAIN at C:\rakudo\install\share\perl6\site\bin\zef line 2
  in block <unit> at C:\rakudo\install\share\perl6\site\bin\zef line 2


C:\rakudo>git diff
diff --git a/src/core/CompUnit/PrecompilationUnit.pm6 b/src/core/CompUnit/PrecompilationUnit.pm6
index 9fb3166..2736853 100644
--- a/src/core/CompUnit/PrecompilationUnit.pm6
+++ b/src/core/CompUnit/PrecompilationUnit.pm6
@@ -57,9 +57,7 @@ role CompUnit::PrecompilationUnit {
             my $source-checksum = $.source-checksum;

             my $srcIO = CompUnit::RepositoryRegistry.file-for-spec($dependency.src) // $dependency.src.IO;
-            unless $srcIO {
-                return False unless $srcIO.e;
-            }
+            return False unless $srcIO and $srcIO.e;
             my $current-source-checksum := nqp::sha1($srcIO.slurp(:enc<iso-8859-1>));
             $RMD(
                 "$.path\nspec: $dependency.spec()\nsource: $srcIO\n"

C:\rakudo>
```

**Comment by zoffixznet** [comment]
Well, I nuked `~/.zef`, `~/.perl6`, and `~/.panda-working`. Then nuked the build dir, compiled from scratch and the problem disappeared.

I think I've used Rakudo Star on this box at some point so perhaps those dirs contained something weird (like, I've not used panda in ages, yet there's .panda-working dir).

The conditional thinko was fixed inhttps://github.com/rakudo/rakudo/commit/9bcc8e523e72e137db10e1b5f01ed9116d33ac2f

Closing without tests.
